File: src/wrappers/go/examples/cgi_snapshot_example.go
// cgi_snapshot_example.go
//
// ITSCAM SDK CGI client example (Go).  Triggers a snapshot.cgi capture,
// writes each resulting image to disk, fetches a lastframe.cgi preview
// and streams MJPEG for 5 seconds.
//
// CGI endpoints are unauthenticated by default on the camera
// (configCgi.blockAPI=false), so --user and --password are optional;
// pass them only when the camera has CGI auth turned on.
//
// Usage:
//
//	go run cgi_snapshot_example.go [--https] [--insecure] \
//	    [--user U --password P] <host>
//
// Copyright (c) 2026 Pumatronix
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"sync/atomic"
	"time"

	"github.com/pumatronix/itscam-sdk-go/itscam"
)

func main() {
	useHttps := flag.Bool("https", false, "use HTTPS instead of HTTP")
	insecure := flag.Bool("insecure", false,
		"skip TLS server certificate verification")
	user := flag.String("user", "",
		"opt-in CGI auth username (omit when blockAPI=false)")
	pass := flag.String("password", "", "opt-in CGI auth password")
	flag.Parse()
	if flag.NArg() < 1 {
		fmt.Fprintln(os.Stderr,
			"Usage: cgi_snapshot_example [--https] [--insecure] "+
				"[--user U --password P] <host>")
		os.Exit(1)
	}
	host := flag.Arg(0)

	scheme := "http"
	port := uint16(80)
	if *useHttps {
		scheme = "https"
		port = 443
	}

	cgi, err := itscam.NewCgiClient()
	if err != nil {
		log.Fatal(err)
	}
	defer cgi.Close()

	if err := cgi.SetBaseUrl(host, port, scheme); err != nil {
		log.Fatal(err)
	}
	if *insecure {
		cgi.SetVerifyServerCertificate(false)
	}
	if *user != "" && *pass != "" {
		if err := cgi.Login(*user, *pass, 10000); err != nil {
			log.Fatalf("login failed: %v", err)
		}
		fmt.Printf("Logged in as %q\n", *user)
	} else {
		fmt.Println("No credentials supplied; talking to the camera " +
			"without CGI authentication.")
	}

	fmt.Println("Fetching lastframe.cgi...")
	last, err := cgi.GetLastFrame(10000)
	if err != nil {
		log.Fatalf("lastframe.cgi failed: %v", err)
	}
	if err := os.WriteFile("lastframe.jpg", last.Data, 0644); err != nil {
		log.Fatal(err)
	}
	fmt.Printf("  -> lastframe.jpg (%s, %d bytes)\n",
		last.MimeType, len(last.Data))

	fmt.Println("Triggering snapshot.cgi (Q=80, mosaic off)...")
	req := itscam.NewSnapshotCgiRequest()
	req.Quality = 80
	imgs, err := cgi.GetSnapshot(req, 15000)
	if err != nil {
		log.Fatalf("snapshot.cgi failed: %v", err)
	}
	fmt.Printf("  -> received %d image(s)\n", len(imgs))
	for i, img := range imgs {
		path := fmt.Sprintf("snapshot-%d.jpg", i)
		if err := os.WriteFile(path, img.Data, 0644); err != nil {
			log.Fatal(err)
		}
		fmt.Printf("     %s: %s, %d bytes\n", path,
			img.MimeType, len(img.Data))
	}

	fmt.Println("Streaming MJPEG for 5 seconds...")
	var frames int64
	if err := cgi.StartMjpegStream(func(f itscam.CgiStreamFrame) {
		n := atomic.AddInt64(&frames, 1)
		if n == 1 {
			_ = os.WriteFile("mjpeg-first.jpg", f.Data, 0644)
		}
	}, 10000); err != nil {
		log.Fatalf("startMjpegStream failed: %v", err)
	}
	time.Sleep(5 * time.Second)
	cgi.StopMjpegStream()
	fmt.Printf("  -> received %d MJPEG frame(s)\n",
		atomic.LoadInt64(&frames))
}

File: src/wrappers/python/examples/cgi_snapshot_example.py
#!/usr/bin/env python3
"""
ITSCAM SDK - CGI snapshot example (Python)

Triggers a snapshot.cgi capture (multi-exposure aware), writes each
resulting image to disk, fetches a lastframe.cgi preview, then streams
5 seconds of MJPEG to demonstrate the callback API.

CGI endpoints are unauthenticated by default on the camera
(``configCgi.blockAPI = false``); pass ``--user`` and ``--password``
only when the camera has CGI auth turned on.

Usage::

    python3 cgi_snapshot_example.py <host> [--https] [--insecure] \\
        [--user U --password P]

Copyright (c) 2026 Pumatronix
"""
from __future__ import annotations

import argparse
import sys
import time

from itscam import ItscamCgiClient, SnapshotCgiRequest


def main() -> int:
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument("host")
    p.add_argument("--user", default=None,
                   help="opt-in CGI auth username (omit when blockAPI=false)")
    p.add_argument("--password", default=None,
                   help="opt-in CGI auth password")
    p.add_argument("--https", action="store_true")
    p.add_argument("--insecure", action="store_true",
                   help="skip TLS server certificate verification")
    args = p.parse_args()

    scheme = "https" if args.https else "http"
    port   = 443 if args.https else 80

    with ItscamCgiClient() as cgi:
        cgi.set_base_url(args.host, port, scheme)
        if args.https and args.insecure:
            cgi.set_verify_server_certificate(False)

        if args.user and args.password:
            cgi.login(args.user, args.password)
            print(f"Logged in as {args.user!r}")
        else:
            print("No credentials supplied; talking to the camera without "
                  "CGI authentication.")

        print("Fetching lastframe.cgi...")
        last = cgi.get_last_frame()
        with open("lastframe.jpg", "wb") as f:
            f.write(last.data)
        print(f"  -> lastframe.jpg ({last.mime_type}, {len(last.data)} bytes)")

        print("Triggering snapshot.cgi (Q=80, mosaic off)...")
        images = cgi.get_snapshot(SnapshotCgiRequest(quality=80))
        print(f"  -> received {len(images)} image(s)")
        for i, img in enumerate(images):
            path = f"snapshot-{i}.jpg"
            with open(path, "wb") as f:
                f.write(img.data)
            print(f"     {path}: {img.mime_type}, {len(img.data)} bytes")

        print("Streaming MJPEG for 5 seconds...")
        frames = []
        def on_frame(frame):
            frames.append(frame.sequence)
            if len(frames) == 1:
                with open("mjpeg-first.jpg", "wb") as f:
                    f.write(frame.data)
        cgi.start_mjpeg_stream(on_frame)
        time.sleep(5)
        cgi.stop_mjpeg_stream()
        print(f"  -> received {len(frames)} MJPEG frame(s)")

    return 0


if __name__ == "__main__":
    sys.exit(main())
